### PR TITLE
add builder for NOMAD

### PR DESCRIPTION
**Pull request by amontoison**
This pull request contains a new build recipe I built using the BinaryBuilder.jl wizard:

* Package name: NOMAD
* Version: v4.0.0

@staticfloat please review and merge.

### CODE REVIEW

File: N/NOMAD/build_tarballs.jl [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+    Linux(:powerpc64le, libc=:glibc),
+    Linux(:i686, libc=:glibc)
+]
+
+
+# The products that we will ensure are always built
+products = [
+    LibraryProduct("libnomad", :libnomad),
+    LibraryProduct("libsgtelib", :libsgtelib),
+    ExecutableProduct("nomad", :nomad; dir_path="nomad.3.9.1/builds/release/bin")
```
Comment (staticfloat) [inline N/NOMAD/build_tarballs.jl]:
```suggestion
    ExecutableProduct("nomad", :nomad, "nomad.3.9.1/builds/release/bin")
```

Comment (staticfloat) [reply N/NOMAD/build_tarballs.jl]: This was a bug in BB, we're fixing it in https://github.com/JuliaPackaging/BinaryBuilder.jl/pull/664

File: N/NOMAD/build_tarballs.jl [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+    "2134e64d5c8728054a797d0067eb49083d66ca64ed8ad7e75c34700b528aeff3",
+]
+
+# Bash recipe for building across all platforms
+script = raw"""
+cd $WORKSPACE/srcdir
+export NOMAD_HOME=${WORKSPACE}/srcdir/nomad.3.9.1
+export PATH=${NOMAD_HOME}/bin:$PATH
+cd $NOMAD_HOME
+./configure
```
Comment (staticfloat) [inline N/NOMAD/build_tarballs.jl]:
```suggestion
./configure --prefix=${prefix} --host=${target}
```

You need to pass some options here, to get it to install into the proper place, to build for the proper target, etc....

Comment (amontoison) [reply N/NOMAD/build_tarballs.jl]: I can only pass the 3 following options with this configure file :
--compiler
--enable-mpi
--help

It's why I move files with a cp command at the end.

**Comment by amontoison** [comment]
I tried to build recipe on MacOS plateform but a hash mismatch : 
```
[ Info: Downloading https://github.com/JuliaPackaging/Yggdrasil/releases/download/PlatformSupport-v2020.2.7/PlatformSupport-x86_64-apple-darwin14.v2020.2.7.x86_64-linux-musl.unpacked.tar.gz to /tmp/jl_Vax7SD-download.gz...
######################################################################## 100,0%
[ Info: No hash cache found
[ Info: Calculated hash 468880dbf10699f8327e7bba4ffeedabfb98a76c6f55c8587c94678005bf46b0 for file /tmp/jl_Vax7SD-download.gz
[ Info: Unpacking /tmp/jl_Vax7SD-download.gz into /home/alexis/.julia/artifacts/jl_1rvi6n...
┌ Error: Tree Hash Mismatch!
│   Expected git-tree-sha1:   b522eb4ee21b2c03c8d885e43f42cd76d9f4fffe
│   Calculated git-tree-sha1: 38fb2513d497b0c8396b1eda7d8248320b70d98a
```
I tried with the last tagged version of BinaryBuilder and its master version. I got the same error.

**Comment by staticfloat** [comment]
You may be using Julia 1.3.1, which has a bug in it.  Please use either Julia 1.3.0, Julia 1.4-rc2 or Julia master.

File: N/NOMAD/build_tarballs.jl (hunk 2 above)
Comment (giordano) [reply N/NOMAD/build_tarballs.jl]: Oh, is this one of those interesting cases where there is a script called `configure` which however is not an autoconf `configure`?

Comment (amontoison) [reply N/NOMAD/build_tarballs.jl]: The content of the configure file is : 

```shell
#!/bin/sh

configure_script=`pwd`/${0##./}
configure_path=${configure_script%/*}
if [ "${NOMAD_HOME%/}" != $configure_path ]; then
    echo "NOMAD_HOME = "$NOMAD_HOME
    echo "NOMAD_HOME environment variable should be set to $configure_path"
    exit 1
fi

echo "NOMAD_HOME is correctly set."

compiler=g++
mpicompiler=mpic++
mpibuild=false
compiler_default=true

makefile_src="$NOMAD_HOME/src/Makefile"

# Makefiles to configure
MAKEFILES_TO_GENERATE="
$NOMAD_HOME/examples/advanced/multi_start/Makefile \
$NOMAD_HOME/examples/advanced/restart/Makefile \
$NOMAD_HOME/examples/advanced/user_search/Makefile \
$NOMAD_HOME/examples/advanced/categorical/single_obj_parallel/Makefile \
$NOMAD_HOME/examples/advanced/categorical/bi_obj/Makefile \
$NOMAD_HOME/examples/advanced/categorical/single_obj/Makefile \
$NOMAD_HOME/examples/advanced/categorical/HeatShield/Makefile \
$NOMAD_HOME/examples/advanced/categorical/bi_obj_parallel/Makefile \
$NOMAD_HOME/examples/advanced/plot/Makefile \
$NOMAD_HOME/examples/interfaces/FORTRAN/example3/Makefile \
$NOMAD_HOME/examples/interfaces/FORTRAN/example2/Makefile \
$NOMAD_HOME/examples/interfaces/FORTRAN/example1/Makefile \
$NOMAD_HOME/examples/interfaces/DLL/bi_obj/Makefile \
$NOMAD_HOME/examples/interfaces/DLL/single_obj/Makefile \
$NOMAD_HOME/examples/interfaces/AMPL/Makefile \
$NOMAD_HOME/examples/basic/library/single_obj_parallel/Makefile \
$NOMAD_HOME/examples/basic/library/bi_obj/Makefile \
$NOMAD_HOME/examples/basic/library/single_obj_sgtelib/Makefile \
$NOMAD_HOME/examples/basic/library/single_obj/Makefile \
$NOMAD_HOME/tools/PSD-MADS_library_mode/Makefile \
$NOMAD_HOME/tools/COOP-MADS/Makefile \
$NOMAD_HOME/tools/SENSITIVITY/cache_inspect/Makefile \
$NOMAD_HOME/tools/SENSITIVITY/problems/styrene/black-box/surrogate/Makefile \
$NOMAD_HOME/tools/SENSITIVITY/problems/styrene/black-box/truth/Makefile \
$NOMAD_HOME/tools/SENSITIVITY/detailed_analysis/Makefile \
$NOMAD_HOME/tools/PSD-MADS/Makefile \
"

for arg in "$@"; do
    case "$arg" in
    --compiler=*)
        compiler=`echo $arg | sed 's/--compiler=//'`
        compiler_default=false
        ;;

    --enable-mpi)
        mpibuild=true;;

    --help)
        echo 'usage: ./configure [options]'
        echo 'options:'
        echo '  --compiler=<compiler>: providing compiler'
        echo '  --enable-mpi: enabling parallel version of NOMAD using MPI'
        echo 'all invalid options exit script but Makefiles are not updated.'
        exit 0
        ;;
    *)
    echo "Unknown argument passed to $0. Run $0 --help for valid options."
    exit 0
    ;;
    esac
done

echo 'Generating makefile ...'


# Clear existing Makefile in src
rm -f $makefile_src

# Verify compilers.
compiler_exists=`which $compiler > /dev/null 2>&1`
if [ $? != 0 ]; then
    echo "$0: Error: compiler $compiler not found"
    if [ $compiler_default == true ]; then
        echo "To specify a compiler, use option --compiler=<compiler>."
    fi
    exit 0
else
    if [ $compiler_default == false ]; then
        echo "$0: Using compiler $compiler."
    fi
fi
# Now do the same with $mpicompiler
if $mpibuild; then
    mpicompiler_exists=`which $mpicompiler >& /dev/null`
    if [ $? != 0 ]; then
        echo "$0: Error: mpi compiler $mpicompiler not found"
        exit 0
    fi
fi

# Note: Continue using $mpicompiler for MPI compilation, even if $compiler was
# specified by the user.
echo "COMPILER = $compiler" >> $makefile_src
echo "COMPILER_MPI = $mpicompiler" >> $makefile_src

# Normal configure, build only non-mpi version.
# Enable MPI: also build mpi version.
if $mpibuild; then
    echo "ALL = nompi mpi" >> $makefile_src
else
    echo "ALL = nompi" >> $makefile_src
fi


if [ -e $makefile_src.in ]; then
    cat $makefile_src.in >> $makefile_src
else
    echo "Error: Expecting file $makefile_src.in to exist"
    exit 1
fi
echo Created $makefile_src.

echo


# Now configure makefiles for examples.
echo 'Generating makefiles for examples ...'

for makefile_to_generate in $MAKEFILES_TO_GENERATE
do
    example_path=${makefile_to_generate%/*}

    # Clear temp file MakefileExample
    rm -f MakefileExample

    # Instructions deduced by the configuration
    echo "COMPILER = $compiler" >> MakefileExample
    echo "COMPILER_MPI = $mpicompiler" >> MakefileExample

    # General compilation instructions for examples
    if [ -e $makefile_to_generate.in ]; then
        cat $makefile_to_generate.in >> MakefileExample
    else
        echo "Warning: Expecting file $makefile_to_generate.in to exist"
    fi

    mv MakefileExample $example_path/Makefile
    echo Created $example_path/Makefile
done

echo 'Configuration complete. Type make to build.'
```

**Comment by amontoison** [comment]
Yes, I use Julia 1.3.1. I will test with Julia 1.4-rc2 and update the pull request afterwards.
Thanks!

**Comment by giordano** [comment]
From the [log of `x86_64-linux-gnu`](https://dev.azure.com/JuliaPackaging/Yggdrasil/_build/results?buildId=1766&view=logs&j=2d1c4fa5-9c55-5514-7e52-41d49a0fb26b&t=d78b8c04-2456-5f80-2fcb-dcfef486d735):
```
==26294==ASan runtime does not come first in initial library list; you should either link runtime to your application or manually preload it with LD_PRELOAD.
┌ Warning: lib/libasan.so cannot be dlopen()'ed
└ @ BinaryBuilder /depot/packages/BinaryBuilder/EUs4B/src/Auditor.jl:143
[...]
ERROR: could not load library "/agent/_work/1/s/N/NOMAD/build/x86_64-linux-gnu-cxx11/WJFzyQX6/destdir/lib/liblsan.so"
/agent/_work/1/s/N/NOMAD/build/x86_64-linux-gnu-cxx11/WJFzyQX6/destdir/lib/liblsan.so: cannot allocate memory in static TLS block
```
These errors are exactly like #336.

**Comment by amontoison** [comment]
Thank you for the Musl patch @giordano !
The build with OpenMP is only available with `x86_64-linux-gnu` and it failed with Yggdrasil. 
I commented the lines associated with that in the source code such that NOMAD is compiled without openMP for all architectures.

When this feature will be also available for Mac and Windows I will try again to build with OpenMP.
Maybe that the problem with Yggdrasil will be solved too.

**Comment by giordano** [comment]
Ok, then I'm going to merge this, thanks!

**Comment by giordano** [comment]
The problems with libgomp are fixed now, in case you want to add it back :wink:

**Comment by amontoison** [comment]
Thank you giordano! I need to add `MPI_jll` as dependency if I want to compile with OpenMP?
I ask the question, because it seems that It worked with only `CompilerSupportLibraries_jll` last time.

By the same token, do you know why NOMAD_jll doesn't work on FreeBSD :  https://cirrus-ci.com/task/5758949131550720 ? It works perfectly on Windows / Mac and Linux :100:

**Comment by giordano** [comment]
> Thank you giordano! Do I need to add MPI_jll as dependency if I want to compile with OpenMP?
> I ask the question, because it seems that It worked with only CompilerSupportLibraries_jll last time.

[MPI](https://en.wikipedia.org/wiki/Message_Passing_Interface) and [OpenMP](https://en.wikipedia.org/wiki/OpenMP) are two different things.  For MPI, we have two different implementations, `MPICH_jll` and `OpenMPI_jll`.  But my understanding is that you need only OpenMP, which is provided by `libgomp` in `CompilerSupportLibraries_jll`.

**Comment by giordano** [comment]
Regarding the FreeBSD issue, I don't know what's the problem :disappointed:

**Comment by amontoison** [comment]
I did a confusion between `OpenMP` and `OpenMPI_jll` :+1: 
I just open a pull request that seeks to compile NOMAD with OpenMP. It was never tested on Windows and Mac :fearful: .

**Pull request merged**